src/main/ruby/core/truffle/cext.rb

```diff
@@ -181,6 +181,17 @@ def fd
   end
 
   class RStringPtr
+    RSTRING_CACHE = {}
+    RSTRING_CACHE_MUTEX = Mutex.new
+
+    private_class_method :new
+
+    def self.get(string)
+      RSTRING_CACHE_MUTEX.synchronize do
+        RSTRING_CACHE.fetch(string.object_id) { |key| RSTRING_CACHE[key] = new(string) }
+      end
+    end
```

Comment:
I'm confused, after TO_NATIVE, it will anyway be the same native pointer referecned by the NativeString, right?
When is this useful? When comparing two `RStringPtr` by identity when they are not yet in native memory?

Reply:
Correct. E.g., this is from the RedCloth gem:

```
char *ts = RSTRING_PTR(str), *te = RSTRING_PTR(str) + RSTRING_LEN(str);
char *t = ts, *t2 = ts;
const char *ch = NULL;
if (te <= ts) return Qnil;
```

`ts` and `te` are both `LLVMAddress` objects, but their values have no relationship to each other. `te` could be less than `ts` or it could be greater than. Logic relying on pointer arithmetic ceases to work.

Reply:
Another way to address this is to just go native on `RSTRING_PTR`, no fancy fat pointer with offsets handling.
Of course, this has the extra cost of copying the bytes to native memory, in the case `RSTRING_PTR` is used but the resulting pointer is never stored itself in native memory.